Java exception handling

cristina-david opened the pull request:
Exceptions are handled by introducing exceptional return variables for program functions. In this PR, a function is added such a return variable if it contains a throw or a function call (as the function being called may escape exceptions). This is too permissive and will be improved in a subsequent PR which uses an uncaught exceptions analysis to get a better approximation of the set of exceptions escaped by a function.

@mgu is currently testing this on datastax so I'm still in the process of fixing reported problems.

peterschrammel commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -47,6 +49,83 @@ static bool is_empty(const goto_programt &goto_program)
+
+        // some handlers may not have been converted (if there was no
+        // exception to be caught); in such a situation we abort
+        for(irept::subt::const_iterator
```
ranged for?

peterschrammel commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -47,6 +49,83 @@ static bool is_empty(const goto_programt &goto_program)
+            return;
+        }
+
+        for(irept::subt::const_iterator
```
ranged for?

peterschrammel commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -302,6 +381,10 @@ void goto_convertt::goto_convert_rec(
   finish_computed_gotos(dest);
   finish_guarded_gotos(dest);
+
+  const symbolt &init_symbol = symbol_table.lookup(INITIALIZE_FUNCTION);
```
cpplint

peterschrammel commented on src/goto-programs/goto_convert_exceptions.cpp:
```diff
@@ -128,6 +128,71 @@ void goto_convertt::convert_msc_leave(
+  const codet &code,
+  goto_programt &dest)
+{
+  assert(code.operands().size()>=1);
```
!empty()?

peterschrammel commented on src/goto-programs/goto_convert_exceptions.cpp:
```diff
@@ -128,6 +128,71 @@ void goto_convertt::convert_msc_leave(
+  }
+
+  // add a SKIP target for the end of everything
+  goto_programt end;
```
this should get a source location

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+Date:   December 2016
+
+\*******************************************************************/
+// #define DEBUG
```
remove

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+
+#include "remove_exceptions.h"
+
+class exceptiont
```
you might want to move this inside remove_exceptionst if it is only used here

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+    stack_catcht stack_catch);
+};
+
+void remove_exceptionst::add_exceptional_returns(
```
function comment block missing

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+       symbol_table.has_symbol(id2string(function_id)+EXC_SUFFIX))
+    {
+      assert(instr_it->code.operands().size()==1);
+      symbolt &function_symbol=symbol_table.lookup(id2string(function_id)+
```
break after =

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+        exc_symbol=exc_symbol.op0();
+
+      // add the assignment with the appropriate cast
+      code_assignt assignment(typecast_exprt(lhs_expr, exc_symbol.type()),
```
cpplint

cristina-david replied:
Not sure what you refer to here. cpplint is not reporting anything.

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+void remove_exceptionst::add_exceptional_returns(
+  goto_functionst::function_mapt::iterator func_it)
+{
+  const irep_idt function_id=func_it->first;
```
&function_id?

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+void remove_exceptionst::replace_throws(
+  goto_functionst::function_mapt::iterator func_it)
+{
+  const irep_idt function_id=func_it->first;
```
&function_id?

peterschrammel commented on src/goto-programs/goto_convert_exceptions.cpp:
```diff
@@ -128,6 +128,71 @@ void goto_convertt::convert_msc_leave(
+  goto_programt::targett catch_instruction=dest.add_instruction();
+  catch_instruction->make_catch();
+  catch_instruction->code.set_statement(ID_catch);
+  catch_instruction->source_location=code.source_location();
```
this should have its `function` set (check also other instructions you generate)

tautschnig replied:
(Just recording that this is done.)

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+      if(symbol_table.has_symbol(id2string(callee_id)+EXC_SUFFIX) &&
+        symbol_table.has_symbol(id2string(caller_id)+EXC_SUFFIX))
+      {
+        symbolt &callee=symbol_table.lookup(id2string(callee_id)+EXC_SUFFIX);
```
make these const

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+
+      if(symbol_table.has_symbol(id2string(function_id)+EXC_SUFFIX))
+      {
+        symbolt &function_symbol=symbol_table.lookup(id2string(function_id)+
```
const?

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+  void add_throw_gotos(
+    goto_functionst::function_mapt::iterator,
+    goto_programt::instructionst::iterator instr_it,
+    stack_catcht stack_catch);
```
Pass a const ref

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+  void add_function_call_gotos(
+    goto_functionst::function_mapt::iterator func_it,
+    goto_programt::instructionst::iterator instr_it,
+    stack_catcht stack_catch);
```
Pass a const ref

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+  goto_programt::targett t_end=goto_program.insert_after(instr_it);
+  goto_programt::targett new_state;
+  // find the end of the function
+  for(goto_programt::instructionst::iterator inner_it=instr_it;
```
Why do you have to iterate to the end? 
Isn't it sufficient to just use new_state as iterator?

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+    for(std::size_t j=stack_catch[i].catch_handlers.size(); j-->0;)
+    {
+      goto_programt::targett new_state_pc;
+      new_state_pc=stack_catch[i].catch_handlers[j].second;
```
make a declaration+assignment

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+  if(symbol_table.has_symbol(id2string(callee_id)+EXC_SUFFIX))
+  {
+    // dynamic dispatch of the escaped exception
+    symbolt &callee_exc_symbol=symbol_table.lookup(id2string(callee_id)+
```
const
break after =

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+      }
+      else // push
+      {
+        exceptiont exception;
```
push the exception onto the stack here

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+        assert(exception_list.size()==instr_it->targets.size());
+
+        // Fill the map with the catch type and the target
+        unsigned i=0;
```
get a ref to stack.back()

cristina-david replied:
Done.

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+        unsigned i=0;
+        for(auto target : instr_it->targets)
+        {
+          exception.catch_handlers.push_back(
```
push_back() on that ref

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+  rr(goto_model.goto_functions);
+}
+
+
```
remove blank lines

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,583 @@
+
+void remove_exceptions(goto_modelt &goto_model)
+{
+  remove_exceptionst rr(goto_model.symbol_table);
```
maybe call it `remove_exceptions`?

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -104,6 +104,13 @@ exprt::operandst java_bytecode_convert_methodt::pop(std::size_t n)
   return operands;
 }
 
+void java_bytecode_convert_methodt::pop_residue(std::size_t n)
```
function comment blocks missing

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -935,6 +961,47 @@ codet java_bytecode_convert_methodt::convert_instructions(
+        catch_handler_expr.set(ID_handler, exc_var);
+        codet catch_handler=code_expressiont(catch_handler_expr);
+
+        std::ostringstream strInt;
```
str_int

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1007,14 +1064,7 @@ codet java_bytecode_convert_methodt::convert_instructions(
-        assertion_failure=true;
-      }
-      const bool use_this(statement!="invokestatic");
+      const bool use_this(statement != "invokestatic");
```
cpplint

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1920,6 +1970,138 @@ codet java_bytecode_convert_methodt::convert_instructions(
+    // add the CATCH-PUSH instruction(s)
+    // be aware of different try-catch blocks with the same starting pc
+    std::size_t pos=0;
+    int end_pc=-1;
```
Is this really an int?

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1920,6 +1970,138 @@ codet java_bytecode_convert_methodt::convert_instructions(
+    // next add the CATCH-POP instructions
+    int start_pc=-1;
+    // check if this is the end of a try block
+    for(std::size_t e=0; e<method.exception_table.size(); e++)
```
maybe a ranged for?

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1920,6 +1970,138 @@ codet java_bytecode_convert_methodt::convert_instructions(
+
+        // create labels for the handlers that match those used
+        // in CATCH-PUSH above
+        std::ostringstream strInt;
```
str_int

peterschrammel commented on src/util/irep_ids.txt:
```diff
@@ -800,4 +802,4 @@ cprover_string_to_char_array_func
 cprover_string_trim_func
-cprover_string_value_of_func
\ No newline at end of file
+cprover_string_value_of_func
```
what has changed here?

cristina-david replied:
Nothing intentional, I probably affected the whitespace.

peterschrammel commented on src/util/irep_ids.txt:
```diff
@@ -738,6 +738,8 @@ java_bytecode_index
 java_instanceof
 java_super_method_call
 java_enum_static_unwind
+push_catch
```
can you put these changes and those to std_code into a separate commit please?

cristina-david replied:
Done.

peterschrammel commented on src/util/irep_ids.txt:
```diff
@@ -740,6 +740,7 @@ java_super_method_call
 java_enum_static_unwind
 push_catch
 handler
+<<<<<<< 33bfd00e8006fbaacd44722426f19a4788d27f1c
```
????????????????

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -7,7 +7,10 @@ Author: Cristina David
 Date:   December 2016
```
Can you squash "Fixed bugs in remove_exceptions: " into previous commits if this is fixing bugs in the same PR?

peterschrammel commented on regression/exceptions/exceptions1/test.desc:
```diff
@@ -3,8 +3,8 @@ test.class
 ^SIGNAL=0$
-^.*assertion at file test.java line 24 function.*: FAILURE$
-\*\* 1 of 31 failed (2 iterations)$
+^.*assertion at file test.java line 26 function.*: FAILURE$
```
If "Updated some regression tests for exceptions " is just fixing a previous commit in the same PR then please squash it

peterschrammel commented on regression/exceptions/exceptions1/test.desc:
```diff
@@ -4,7 +4,7 @@ test.class
 ^SIGNAL=0$
 ^.*assertion at file test.java line 26 function.*: FAILURE$
-\*\* 1 of 35 failed (2 iterations)$
+\*\* 1 of 4 failed (2 iterations)$
```
Please squash the changes of "Addressed reviewers comments" into the corresponding commits.

cristina-david commented:
@peterschrammel I've addressed your comments. Regarding the source location for CATCH push and pop, I've added them where you want them but, in general, I didn't try hard because I'm removing those instructions later on. If you really want the there, I have to also check the conversion to codet.

tautschnig commented:
Shouldn't the regression tests just go in the cbmc-java folder?

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,590 @@
+  Forall_goto_program_instructions(instr_it, goto_program)
+  {
+    // is this a handler
+    if(instr_it->type==CATCH && instr_it->code.find(ID_handler)!=get_nil_irep())
```
Use `instr_it->code.find(ID_handler).is_not_nil()`

cristina-david commented:
@tautschnig Thanks for your comments, I've implemented them.

tautschnig commented on src/symex/symex_parse_options.cpp:
```diff
@@ -364,7 +364,6 @@ bool symex_parse_optionst::process_goto_program(const optionst &options)
     remove_vector(goto_model);
     // Java virtual functions -> explicit dispatch tables:
     remove_virtual_functions(goto_model);
-    // Java instanceof -> clsid comparison:
```
I think that bit might still be missing: it seems only cbmc_parse_options.cpp has received the new exceptions processing, while symex, goto-analyzer, goto-instrument would still benefit from it.

cristina-david replied:
@tautschnig The exception processing is now called in symex, goto-analyzer and goto-instrument (I've actually introduced the call in symex_parse_options.cpp after your initial comments but was slightly misplaced).

tautschnig commented:
@cristina-david Thanks a lot for all the changes! @peterschrammel Would you mind reviewing your decision?

Overall we've got three unrelated problems that presently make all continuous integration attempts fail: 1) Unrelated linter warnings (cf #376). 2) A regression test that is too specific on instruction numbers - this should be fixed, but it would be great to have #505 merged first. 3) Limited time budget on AppVeyor.

peterschrammel commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,586 @@
+         corresponding exception handlers
+
+\*******************************************************************/
+void remove_exceptionst::add_throw_gotos(
```
Polishing: there should be a blank line separating the comment block and the function.

peterschrammel commented:
I'd still fix the failing test as part of this pr if it caused failing it.

tautschnig commented:
I do agree with fixing the failing test, but as that fix will involve use of regular expressions a prior merge of #505 would be very helpful.

peterschrammel commented:
Can be rebased now.

cristina-david commented:
@peterschrammel This has been cleaned up: all cbmc-java tests pass, no more use of the mode of the entry function, the mode check (==ID_java) has been eliminated from goto_convert, the named operands have been replaced by positional ones in java_bytecode_convert in order not to trip the live range computations, etc.

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+      lhs_expr_null.type()=typet(ID_pointer, empty_typet());
+
+      exprt rhs_expr_null;
+      rhs_expr_null=null_pointer_exprt(pointer_typet());
```
Isn't the `pointer_typet` missing a subtype here?

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+      t_null->source_location=instr_it->source_location;
+      t_null->code=code_assignt(
+        lhs_expr_null,
+        typecast_exprt(rhs_expr_null, lhs_expr_null.type()));
```
If the subtype is set properly as suggested by my above comment, then you shouldn't even need any typecast here.

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+        lhs_expr_null,
+        typecast_exprt(rhs_expr_null, lhs_expr_null.type()));
+      t_null->function=instr_it->function;
+      return;
```
So we just do this once, before the first function call or throw instruction. What about code like
```
if(foo)
{
  bar();
}
foobar();
```
which would have the pointer assignment inserted before the call to `bar`, and thus within a branch. If that branch isn't taken, but `foobar` throws an exception, then the pointer assignment has never been seen?!

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+      // replace "throw x;" by "f#exception_value=x;"
+      symbol_exprt lhs_expr;
+      lhs_expr.set_identifier(id2string(function_id)+EXC_SUFFIX);
+      lhs_expr.type()=function_symbol.type;
```
This type seems different from the `void*` that is being generated in `add_exceptional_returns`?!

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+      // find the symbol corresponding to the thrown exceptions
+      exprt exc_symbol=instr_it->code;
+      while(exc_symbol.id()!=ID_symbol && exc_symbol.has_operands())
+        exc_symbol=exc_symbol.op0();
```
Why is it necessary to look for operands?

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+                              exc_symbol);
+      // now turn the `throw' into `assignment'
+      instr_it->type=ASSIGN;
+      instr_it->code=assignment;
```
Wouldn't that also require a goto towards code executing the destructors, and then leave the function?

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+
+      // can exceptions escape?
+      if(symbol_table.has_symbol(id2string(callee_id)+EXC_SUFFIX) &&
+        symbol_table.has_symbol(id2string(caller_id)+EXC_SUFFIX))
```
Nit picking: indent line by one more space

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+
+        symbol_exprt rhs_expr;
+        rhs_expr.set_identifier(id2string(callee_id)+EXC_SUFFIX);
+        rhs_expr.type()=callee.type;
```
See above: `symbol_exprt rhs_expr=callee.symbol_expr();`

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+
+        symbol_exprt lhs_expr;
+        lhs_expr.set_identifier(id2string(caller_id)+EXC_SUFFIX);
+        lhs_expr.type()=caller.type;
```
`symbol_exprt lhs_expr=caller.symbol_expr();`

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+      if(symbol_table.has_symbol(id2string(function_id)+EXC_SUFFIX))
+      {
+        const symbolt &function_symbol=
+          symbol_table.lookup(id2string(function_id)+EXC_SUFFIX);
```
Just a thought: You can do `has_symbol` and `lookup` in a single step by using the 2-argument version of `lookup` that takes a reference-to-a-pointer as second argument.

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+        lhs_expr_null.set_identifier(id2string(function_id)+EXC_SUFFIX);
+        lhs_expr_null.type()=function_symbol.type;
+        exprt rhs_expr_null;
+        rhs_expr_null=null_pointer_exprt(pointer_typet());
```
Various comments from above apply: use `symbolt::symbol_expr()`, set the subtype of `pointer_typet`

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+        t_null->source_location=instr_it->source_location;
+        t_null->code=code_assignt(
+          lhs_expr_null,
+          typecast_exprt(rhs_expr_null, lhs_expr_null.type()));
```
Again, I'm not sure this typecast is required.

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+        // add the assignment exc=f#exception_value
+        symbol_exprt rhs_expr_exc;
+        rhs_expr_exc.set_identifier(id2string(function_id)+EXC_SUFFIX);
+        rhs_expr_exc.type()=function_symbol.type;
```
See above.

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+
+  // find the end of the function
+  goto_programt::targett new_state;
+  for(new_state=instr_it; !new_state->is_end_function(); new_state++) {}
```
You should be able to find that as `goto_program.instructions.end()--;` I believe. (Possibly to combine with an assertion that this indeed is `is_end_function()`

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+  assert(instr_it->code.operands().size()==1);
+
+  // make sure that these are always forward gotos
+  assert(!instr_it->is_backwards_goto());
```
I thought it was a THROW?

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+  assert(!instr_it->is_backwards_goto());
+
+  // find the end of the function
+  goto_programt::targett new_state;
```
Nit picking: is that really a "state"?

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+      // find the symbol corresponding to the caught exceptions
+      exprt exc_symbol=instr_it->code;
+      while(exc_symbol.id()!=ID_symbol)
+        exc_symbol=exc_symbol.op0();
```
Same question as above (well, also same code -> refactor?): why does one need to iterate here?

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+    goto_programt::targett t_end=goto_program.insert_after(instr_it);
+    t_end->make_goto(new_state);
+    t_end->source_location=instr_it->source_location;
+    t_end->function=instr_it->function;
```
Question as before: what about destructors that need to be invoked?

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+
+    symbol_exprt callee_exc;
+    callee_exc.set_identifier(id2string(callee_id)+EXC_SUFFIX);
+    callee_exc.type()=callee_exc_symbol.type;
```
See above (`symbolt::symbol_expr()`)

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+
+    // add a null check (so that instanceof can ba applied)
+    equal_exprt eq_null(typecast_exprt(callee_exc, pointer_typet()),
+                        null_pointer_exprt(pointer_typet()));
```
As before: `pointer_typet` without a subtype.

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+      }
+    }
+
+    // add a null check (so that instanceof can ba applied)
```
Type: ba -> be

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+    t_null->make_goto(next_it);
+    t_null->source_location=instr_it->source_location;
+    t_null->function=instr_it->function;
+    t_null->guard=eq_null;
```
So what if none of the handlers applied?

cristina-david replied:
Now, after the conditional GOTOs to the exception handlers, there is an unconditional GOTO to the end of the function.

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,598 @@
+    instrument_exception_handlers(it);
+    add_gotos(it);
+    instrument_function_calls(it);
+    replace_throws(it);
```
I don't think I fully understand the prescribed sequence of actions here. It seems that `add_gotos` handles multiple kinds of instructions, whereas the others each iterate over the full goto program to just find and process a single kind of instructions. Is that necessary?

cristina-david replied:
I've restructured the code such that we only have two passes over the whole program: the first one adds the necessary exceptional return variables and the second one (see method instrument_exceptions) does all the rest of the instrumentation.

tautschnig commented on src/goto-programs/remove_exceptions.h:
```diff
@@ -0,0 +1,24 @@
+
+#include <goto-programs/goto_model.h>
+
+#define EXC_SUFFIX "#exception_value"
```
Does this `#define` have to go in the header file?

cristina-david replied:
This might be needed outside remove_exceptions.cpp. For instance, I use it in java_entry_point.cpp in PR #578.

tautschnig replied:
Ok!

tautschnig commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -947,21 +1029,30 @@ codet java_bytecode_convert_methodt::convert_instructions(
-        // if this athrow is generated by an assertion, then skip it
-        c=code_skipt();
-        assertion_failure=false;
+        // TODO throw NullPointerException instead
```
Is this still a "TODO" - I thought you've now implemented the fix?

cristina-david replied:
This is a TODO for when we will replace all the instrumentation related to runtime exceptions, which currently consists of adding assertions, by throwing exceptions. There are several such TODOs in java_bytecode_convert_method.cpp which we can only address once the exception handling works.

tautschnig commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1920,6 +2004,113 @@ codet java_bytecode_convert_methodt::convert_instructions(
+    // must track it separately
+    bool first_handler=false;
+    // check if this is the end of a try block
+    for(auto exception_row : method.exception_table)
```
That should probably be `const auto &exception_row`

tautschnig commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1931,6 +2122,18 @@ codet java_bytecode_convert_methodt::convert_instructions(
+      // we don't worry about exception handlers as we don't load the
+      // operands from the stack anyway -- we keep explicit global
+      // exception variables
+      for(auto exception_row : method.exception_table)
```
`const auto &exception_row`

forejtv commented:
@cristina-david what is the status of this? Did I understand correctly in Monday's meeting that you consider this PR complete and reviewable, except for one point (I only remember the keyword "destructors") which you planned to do in a separate PR?

tautschnig commented:
I think most of my comments have been addressed, except for a few that might just be resolved by responding to them here on github.

peterschrammel commented:
I've discussed the remaining issues with @cristina-david yesterday. A bit of redesign is necessary to resolve the scoping properly (DEAD instructions). Cristina is currently working on that.

cristina-david commented:
The problem is related to Michael's comments regarding the invocation of destructors. In the current design, I am translating the throws to gotos after the dead statements have been introduced. Thus, the placement of the dead statements doesn't take into consideration the throws as all, which is bad. I'm currently doing some redesigning to fix this. @tautschnig I'll answer the rest of your questions once this is done. Depending on how much I need to change, some might become obsolete. @forejtv this is related to one of the tickets we've written on Monday.

cristina-david commented:
My initial go at doing the exception handling instrumentation before the DEAD are being inserted in goto_convert proved unfeasible due to function calls, which may or may not throw. Essentially, it's very difficult to have a clear image of the CFG at goto_convert level when having to deal with exceptions. Thus, I backtracked to recreating the DEAD statements on the goto-program at the same time with doing the exception handling instrumentation. This seems to work so that's good. However, there is a remaining problem, the goto basic block structure doesn't correctly consider try-catch, resulting in some of the variable declarations being placed outside their corresponding try-catch blocks (and thus messing up the DEAD replacement).

````
void foo(int y)
    {
      try{
	int a_outer=1;
	a_outer++;
	try {
	  int x=0;
....
````   

In ```foo```, ````a_outer```` and ````x```` are declared in the outer and inner try-catch, respectively. However, the goto basic block structure doesn't correctly reflect this:

````
pc0:
  {
    int a_outer;

  pc0$:
    {
      int x;

    pc0$$:
      {
        {
          irep("(\"side_effect\" \"statement\" (\"push_catch\") \"label\" (\"label\" \"\" (\"pc41\")) \"exception_list\" (\"exception_list\" \"\" (\"java::A\")))");
          ;
        }
        a_outer = 1;
        a_outer = a_outer + (int)1;
        {
          irep("(\"side_effect\" \"statement\" (\"push_catch\") \"label\" (\"label\" \"\" (\"pc32\")) \"exception_list\" (\"exception_list\" \"\" (\"java::A\")))");
          ;
        }
        x = 0;
````
I'm currently looking into forcing the try-catch to start a new basic block.

cristina-david commented:
@peterschrammel @tautschnig This now adds the required DEAD instructions + addresses the other comments.

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,554 @@
+    // initialize the exceptional return with NULL
+    symbol_exprt lhs_expr_null=new_symbol.symbol_expr();
+    exprt rhs_expr_null;
+    rhs_expr_null=null_pointer_exprt(pointer_typet(empty_typet()));
```
How about `null_pointer_exprt rhs_expr_null(pointer_typet(empty_typet()));`

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,554 @@
+    // next we reset the exceptional return to NULL
+    symbol_exprt lhs_expr_null=function_symbol.symbol_expr();
+    exprt rhs_expr_null;
+    rhs_expr_null=null_pointer_exprt(pointer_typet(empty_typet()));
```
`null_pointer_exprt rhs_expr_null(pointer_typet(empty_typet()));`

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,554 @@
+  assert(instr_it->type==CATCH && instr_it->code.has_operands());
+
+  // retrieve the exception variable
+  const irept &exception=instr_it->code.op0();
```
`const exprt &exception` would avoid the static cast later on.

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,554 @@
+  goto_programt::targett end_function;
+  for(end_function=instr_it;
+      !end_function->is_end_function();
+      end_function++) {}
```
The loop-free version:
```
assert(!goto_program.empty());
goto_programt::targett end_function=--goto_program.instructions.end();
assert(end_function->is_end_function());
```

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,554 @@
+    goto_programt::targett end_function;
+    for(end_function=instr_it;
+        !end_function->is_end_function();
+        end_function++) {}
```
See above; it might actually be worthwhile adding a method `get_end_function` to `goto_program_templatet`

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,554 @@
+  stack_catcht stack_catch; // stack of try-catch blocks
+  std::vector<std::vector<exprt>> stack_locals; // stack of local vars
+  std::vector<exprt> locals;
+//  bool skip_dead=false;
```
What would this be good for (plus the comment has a surprising non-indent)

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -0,0 +1,554 @@
+    }
+  }
+}
+
```
A single blank line will suffice

tautschnig commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -5,7 +5,7 @@ Module: JAVA Bytecode Language Conversion
 
 \*******************************************************************/
-
+#define DEBUG
```
Should this be in the final version?

tautschnig commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -989,6 +1029,48 @@ codet java_bytecode_convert_methodt::convert_instructions(
 
+    // we throw away the first statement in an exception handler
+    // as we don't know if a function call had a normal or exceptional return
+    size_t e;
```
Use an iterator instead of the index (and then compare the iterator to end() after that loop)

tautschnig commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -989,6 +1029,48 @@ codet java_bytecode_convert_methodt::convert_instructions(
+        catch_handler_expr.get_sub().resize(1);
+        catch_handler_expr.get_sub()[0]=exc_var;
+
+        codet catch_handler=code_expressiont(catch_handler_expr);
```
`code_expressiont catch_handler(catch_handler_expr);`

tautschnig commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1992,6 +2076,115 @@ codet java_bytecode_convert_methodt::convert_instructions(
+
+    // be aware of different try-catch blocks with the same starting pc
+    std::size_t pos=0;
+    size_t end_pc=0;
```
Nit: use `std::size_t` everywhere instead of mixing in plain `size_t`

tautschnig commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1992,6 +2076,115 @@ codet java_bytecode_convert_methodt::convert_instructions(
+      // add the CATCH-POP before the end of the try block
+      if(cur_pc<exception_row.end_pc &&
+        !working_set.empty() &&
+        *working_set.begin()==exception_row.end_pc)
```
Nit picking: Indent the two lines above by 1 more space

tautschnig commented on src/util/std_code.h:
```diff
@@ -1122,6 +1122,35 @@ inline const side_effect_expr_throwt &to_side_effect_expr_throw(
+  inline side_effect_expr_catcht():side_effect_exprt(ID_push_catch)
+  {
+  }
+  inline explicit side_effect_expr_catcht(const irept &exception_list):
```
No need to mark functions `inline` within the class definition.

cristina-david commented:
@peterschrammel and @tautschnig I've addressed the comments.
